src/Compilers/CSharp/Portable/BoundTree/BoundDagEvaluation.cs

```diff
@@ -10,12 +10,20 @@ namespace Microsoft.CodeAnalysis.CSharp
 {
     partial class BoundDagEvaluation
     {
-        public override bool Equals([NotNullWhen(true)] object? obj) => obj is BoundDagEvaluation other && this.Equals(other);
-        public virtual bool Equals(BoundDagEvaluation other)
+        public sealed override bool Equals([NotNullWhen(true)] object? obj) => obj is BoundDagEvaluation other && this.Equals(other);
+        public bool Equals(BoundDagEvaluation other)
         {
             return this == other ||
-                this.Kind == other.Kind &&
-                this.Input.Equals(other.Input) &&
+                this.IsEquivalentTo(other) &&
+                this.Input.Equals(other.Input);
+        }
+
+        /// <summary>
+        /// Check if this is equivalent to the <paramref name="other"/> node, ignoring the input.
+        /// </summary>
+        public virtual bool IsEquivalentTo(BoundDagEvaluation other)
+        {
+            return this.Kind == other.Kind &&
```

Comment:
Why remove `this == other ||` from `IsEquivalentTo`? Do we just never hit that path?

Reply:
> Do we just never hit that path?

Right. We only call `IsEquivalentTo` when `this == other` has failed.

Note we've split the original method in two, moving the ref comparison to the base so we don't need it in the subtypes either.

Reply:
Consider adding assertions to that effect (references are not equal) or keeping the reference equality test, in case we use `IsEquivalentTo` in more scenarios in the future.

Reply:
A closer look at `BoundDagTemp.Equals` shows that we could actually hit this when only `BoundDagTemp.Index` is different (used for deconstruction node where we have multiple temps per evaluation).

I will add it back.